### ISSUE TITLE
Parts of Speech Tagger

### ISSUE THREAD

**Issue by alanjames1987** [NaturalNode/natural#117]
Has the parts of speech tagger been ported from nltools?

**Comment by silentrob** [comment]
Well done. This would be a great addition to the library. It would be nice if the actual tags are abstracted as well. Im a fan of PennBank Tags[1], but I could see other people wanting to use Twitter or something else.

1. http://www.ling.upenn.edu/courses/Fall_2003/ling001/penn_treebank_pos.html

**Comment by Hugo-ter-Doest** [comment]
The module already abstracts from the actual tags. You can use any tags you like in lexicon and rules. See the examples for Dutch and English in de data folder.

Regards,
Hugo

**Comment by Hugo-ter-Doest** [comment]
I think this one can be closed as I merged it into natural.

**Comment by kkoch986** [comment]
I agree, thanks!

**Comment by eqyiel** [comment]
@Hugo-ter-Doest where can I get the files required for BrillPOSTagger?  I thought they might be `lexicon.js` and `BrillTransformationRules.js` respectively from [pos-js](https://github.com/dariusk/pos-js) but it doesn't seem to work:

```
> process.cwd()
'/home/eqyiel/dev/gentlebot/src'
> fs.lstatSync('../node_modules/parts-of-speech')
{ dev: 65027,
  mode: 16877,
  nlink: 3,
  uid: 1000,
  gid: 100,
  rdev: 0,
  blksize: 4096,
  ino: 6043083,
  size: 4096,
  blocks: 8,
  atime: Sun Apr 17 2016 10:53:43 GMT+0930 (ACST),
  mtime: Sun Apr 17 2016 10:53:43 GMT+0930 (ACST),
  ctime: Sun Apr 17 2016 10:53:43 GMT+0930 (ACST),
  birthtime: Sun Apr 17 2016 10:53:43 GMT+0930 (ACST) }
> var tagger = new natural.BrillPOSTagger('./../node_modules/parts-of-speech/lexicon.js', './../node_modules/parts-of-speech/lexer.js', 'N', (error) => {
  if (error) {
  console.log(error);
  } else {
  var sentence = ["I", "see", "the", "man", "with", "the", "telescope"];
  console.log(JSON.stringify(tagger.tag(sentence)))
  }
})
var tagger = new natural.BrillPOSTagger('./../node_modules/parts-of-speech/lexicon.js', './../node_modules/parts-of> var tagger = new natural.BrillPOSTagger('./../node_modules/parts-of-speech/lexicon.js', './../node_modules/parts-of- speech/lexer.js', 'N', (error) => {
...   if (error) {
.....   console.log(error);
.....   } else {
.....   var sentence = ["I", "see", "the", "man", "with", "the", "telescope"];
.....   console.log(JSON.stringify(tagger.tag(sentence)))
.....   }
... })
undefined
> SyntaxError: Expected " ", "*", "//", "\n", "\r", "\r\n", "\t" or [a-zA-Z_0-9_\-.,()] but "/" found.
> 
```

Thanks for your work on natural!

**Comment by Hugo-ter-Doest** [comment]
To create a Brill tagger you need to specify a file with a lexicon and a file with a set of transformation rules.  Your output shows these files:
'./../node_modules/parts-of-speech/lexicon.js', './../node_modules/parts-of- speech/lexer.js',
I do not recognise these files. Are these from another module?

You can use  this file for the lexicon:
natural/lib/natural/brill_pos_tagger/data/English/lexicon_from_posjs.json
and this file for the transformation rules:
natural/lib/natural/brill_pos_tagger/data/English/tr_from_posjs.txt

Best regards, Hugo

**Comment by arjunskumar47** [comment]
@Hugo-ter-Doest 

The `natural/lib/natural/brill_pos_tagger/data/English/lexicon_from_posjs.json` file itself is 4mb in size. 

Since the file is static can we consider caching the file for consecutive use?

**Comment by Hugo-ter-Doest** [comment]
Not sure if I understand what you mean. The lexicon is read in memory when the tagger is created.

Hugo

**Comment by arjunskumar47** [comment]
My concern is around using `fs.readFile` every time I use something like in my code.

```
import { BrillPOSTagger } from 'natural'
let tagger = new BrillPOSTagger(lexiconFilePath, rulesFilePath, 'N', (err) => {
  // do something
});
```

**Comment by Hugo-ter-Doest** [comment]
I see what you mean. Creating a tagger is asynchronous because reading the lexicon and the transformation rules is done asynchronously and this propagates to the constructor. I could replace the asynchronous readFile with synchronous calls to readFileSync and make BrillPOSTagger synchronous. Although this is not really how nodejs was meant, it would make things simpler.

Or maybe you have another idea?

Regards, 
Hugo

**Comment by arjunskumar47** [comment]
That's fine, but I think it'll be continuous I/O for a very frequent initialisation for `BrillPOSTagger` ( eg: typeahead ).

I was thinking in the lines of an LRU or file cache.

**Comment by Hugo-ter-Doest** [comment]
Why do you need frequent initialisation? You can initialise the tagger once and feed all the strings you like to the tagger.

**Comment by arjunskumar47** [comment]
@Hugo-ter-Doest In my case what I'm doing is for a particular API request ( for typeahead ), I'm using the tagger, which can be very frequent after debouncing.

And initializing once and using every time isn't a good idea in this context or otherwise as it'll hold some memory which won't be freed on GC

**Comment by Hugo-ter-Doest** [comment]
I think it is not a good idea to integrate a caching solution into the tagger as it serves one (your) specific application.

I am considering taking the file I/O out of the tagger. This would imply that you need to pass a lexicon (object or function) and a set of transformation rules to the tagger instead of passing filenames. In that way you can organise the lexicon the way you want to. Same for transformation rules.

Creating a tagger would then look like:

var lexicon = new Lexicon(lexiconFilename);
var rules = new Ruleset(rulesFilename);
var tagger = new(lexicon, rules, defaultCategory);

You could provide your own Lexicon object as long as it provides the right methods. Your own Lexicon could be smart in caching the lexicon file in memory.

Regards,
Hugo

**Comment by arjunskumar47** [comment]
Hugo,

That sounds great as I can even provide a file from a URL.

**Comment by Hugo-ter-Doest** [comment]
Ok, I will give it a try tonight.

Hugo

**Comment by Hugo-ter-Doest** [comment]
I refactored the tagger so that creating a rule set and a lexicon from files is separate from creating the tagger. My fork is here:
https://github.com/Hugo-ter-Doest/natural
The tests run ok.

I need to clean up the code before I create a pull request. Will do that tomorrow.

Creating a tagger is now done as follows:
var lexicon = new Lexicon(lexiconFilename, defaultCategory);
var rules = new Ruleset(rulesFilename);
var tagger = new BrillPOSTagger(lexicon, rules);

Regards,
Hugo

**Comment by Hugo-ter-Doest** [comment]
Created a pull request:
[https://github.com/NaturalNode/natural/pull/318]

Hugo

**Comment by kkoch986** [comment]
merged #318
